### PR TITLE
Kmer index allow span ref range

### PR DESCRIPTION
**Pull request by zrm22**
## Description

Updating index code to allow for Kmers come from different reference ranges.  This does not handle the mapping side of things which will be done in a future pull request.

## Type of change

_What type of changes does your code introduce? Put an `x` in boxes that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated relevant documentation

## Changelog entry

_Please add a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
Updating Build Kmer Index to allow Kmers to be from multiple reference ranges.
```

### CODE REVIEW

File: src/main/kotlin/net/maizegenetics/phgv2/pathing/AlignmentUtils.kt [hunk 1]
```diff
@@ -22,7 +23,10 @@ import kotlin.math.min
  * data class to hold the KmerMap information.  From the PHGv1 source code:
  */
 //data class KmerMapData(val haplotypeListId: Int, val rangeToBitSetMap: Map<Int, BitSet>, val kmerHashToLongMap: Long2LongOpenHashMap)
-data class KmerMapData(val rangeToBitSetMap: Map<ReferenceRange, BitSet>, val kmerHashToLongMap: Long2LongOpenHashMap)
+//data class KmerMapData(val rangeToBitSetMap: Map<ReferenceRange, BitSet>, val kmerHashToLongMap: Long2LongOpenHashMap)
```
Comment (tcasstevens) [inline src/main/kotlin/net/maizegenetics/phgv2/pathing/AlignmentUtils.kt:26]: Remove this comment?

File: src/test/kotlin/net/maizegenetics/phgv2/pathing/MapKmersTest.kt [hunk 2]
```diff
@@ -252,27 +253,91 @@ class MapKmersTest {
...
+            if(unSeenHapIds.isNotEmpty()) {
+                println("Error:")
+                println("Kmer: $kmerHashLong")
+                println("Unseen HapIds: $unSeenHapIds")
             }
+            assertTrue(unSeenHapIds.isEmpty())
+
         }
+
+//        //loop through the kmers in kmerHashToLong Map
```
Comment (tcasstevens) [inline src/test/kotlin/net/maizegenetics/phgv2/pathing/MapKmersTest.kt:299]: Remove these commented lines?

File: src/test/kotlin/net/maizegenetics/phgv2/pathing/MapKmersTest.kt [hunk 3]
```diff
@@ -466,7 +531,8 @@ class MapKmersTest {
 
         val rangeToHapidIndexMap = mapOf(1 to mapOf("1" to 0, "2" to 1), 2 to mapOf("3" to 0, "4" to 1))
 
-        val kmerHashOffsetMap = Long2LongOpenHashMap()
+//        val kmerHashOffsetMap = Long2LongOpenHashMap()
```
Comment (tcasstevens) [inline src/test/kotlin/net/maizegenetics/phgv2/pathing/MapKmersTest.kt:534]: Remove this and following commented lines?

**Review by tcasstevens — APPROVED** [review]

File: src/main/kotlin/net/maizegenetics/phgv2/pathing/BuildKmerIndex.kt [hunk 4]
```diff
@@ -138,6 +138,8 @@ class BuildKmerIndex: CliktCommand(help="Create a kmer index for a HaplotypeGrap
         val refRangeToIndexMap = graph.refRangeToIndexMap()
         val hapidToRefrangeMap = graph.hapIdToRefRangeMap()
 
+        val maxHapsToKeep = sampleGametes.size * 2
```
Comment (lynnjo) [inline src/main/kotlin/net/maizegenetics/phgv2/pathing/BuildKmerIndex.kt:141]: The header comment at the top of BuildKmerIndex.kt says "Only hashes observed in exactly one reference range will be kept. ...".  WIll you update this comment to reflect the changes you are making here?

File: src/test/kotlin/net/maizegenetics/phgv2/pathing/MapKmersTest.kt [hunk 5]
```diff
@@ -252,26 +253,46 @@ class MapKmersTest {
...
-            val (refRangeId, offset) = AlignmentUtils.decodeRangeIdAndOffset(encodedOffset)
-            //get the ref range so we can look up refRange -> HapId Index
-            val hapIdToIndexMap = refRangeIdToHapIdMap[refRangeId]!!
-
+            //Loop through the values and test the kmers keeping track of the hapIds which are not in any known refRange.
             val hapIds = kmerMapToHapids[kmerHashLong]!!
+            val unSeenHapIds = mutableSetOf<String>()
+            unSeenHapIds.addAll(hapIds)
+
+            println("Kmer: $kmerHashLong")
```
Comment (lynnjo) [inline src/test/kotlin/net/maizegenetics/phgv2/pathing/MapKmersTest.kt:266]: Might want to remove (or leave for debugging but comment out) this line if it going to be printed alot.  Could be distracting in the log file

File: src/test/kotlin/net/maizegenetics/phgv2/pathing/MapKmersTest.kt [hunk 6]
```diff
@@ -252,26 +253,46 @@ class MapKmersTest {
...
+            for((refRange,encodedOffset) in values) {
+                val (refRangeId, offset) = AlignmentUtils.decodeRangeIdAndOffset(encodedOffset)
+                println("\t$refRange $refRangeId $offset")
+                //get the ref range so we can look up refRange -> HapId Index
+                val hapIdToIndexMap = refRangeIdToHapIdMap[refRangeId]!!
+
+                val hapIds = kmerMapToHapids[kmerHashLong]!!
+                println("\t\t$hapIds")
```
Comment (lynnjo) [inline src/test/kotlin/net/maizegenetics/phgv2/pathing/MapKmersTest.kt:274]: Remove?

**Review by lynnjo — APPROVED** [review]

**Pull request merged**